### PR TITLE
REALMC-7339 Add table printing to terminal

### PR DESCRIPTION
**Pull request by jo-tse-mongo**

### CODE REVIEW

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:54]: move this to the top of the cases, remove the case body (remove `val = ""` below), and add an inline comment here:

```go
case nil: // leave zero-value
```

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:50]: lets remove this `case` until we need it (feel free to let me know I'm missing something and we need it now)

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:49]: lets avoid any transformation logic for now and just set the raw `string` value (feel free to let me know I'm missing something and we need this now)

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:60]: [nit]

```go
if v := len(normalizedValue); v > t.columnWidths[key] {
  t.columnWidths[key] = v
}
```

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:77]: [nit] `tempBuilder` => `builder`

also, do you think we could accomplish this with just one string Builder?  it seems like we ultimately want:

```go
h := color.New(color.Bold).SprintFunc()(header)
builder.WriteString(fmt.Sprintf("%s"+strings.Repeat(" ", spaceCount), h))
```

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:79]: [nit] `for _, header := range t.headers`

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:81]: can you run me through this logic real quick?  having trouble figuring out why we're subtracting the gutter from the header width

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:23]: [nit] `columnBreakWidth` => `gutterWidth`

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:97]: same thing here, i'll let you try to work through this implementation where we're only needing a single `var builder strings.Builder`

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:99]: same naming scheme here: `data` has many `row`s, `row` is build of `key` and `value`s, etc

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:122]: same thing here, i'll let you try to work through this implementation where we're only needing a single `var builder strings.Builder`

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:135]: prefix this with `logField` (and for `data` below)

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:139]: to satisfy this interface, the return signature here will want to be `(string, error)`
also, at the start of this and `Payload()` below, we can add a `if err := t.validate(); err != nil { return "", err }` check.  inside that validate check is where you can move the logic you had for empty/nil headers

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:148]: just a reminder to use the `const`s you defined for these values here (`logFieldHeaders` and `logFieldData`)

**Review by makesitgo — COMMENTED** [review]
did a first pass through the main `table.go` file.  really no major comments from me though, a lot of them are just some standard go practices (naming, organization and such) and other various nits of mine.  this is lookin really good so far!

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:50]: the float gets cut off at a certain number, so i make it formatted by %g (which handles long floats and defaults to %f if it's short)

Comment (makesitgo) [reply internal/terminal/table.go:50]: neat!  i definitely did not know about the `%g` formatter... i guess my question here is: what `float`s are already trying to print here?  if this was just for future-proofing, then I'll still suggest we remove this `case` until the need/requirement for it shows up

**Comment by tsedgwick** [comment]
before merging, please remember to update the PR title in the form of `REALMC-xxxx: imperative sentence`

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:26]: I return a pointer so that I could return a nil instead of a table if the values are empty (header, data) 
I also have a question in regards to the other comment that involves implementing a validate function, since if I return a nil here, there should be no real way for them to create a table without going through the initial validation

Comment (makesitgo) [reply internal/terminal/table.go:26]: yeah let's avoid using a pointer just to return `nil` here, the zero-value of the struct will be just fine

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:49]: If we don't replace any possible newlines, it breaks up the table formatting (\t, \n, etc) so I try to replace them here

Comment (makesitgo) [reply internal/terminal/table.go:49]: i see, that makes a lot of sense.  the one thing is I'm not sure we have/expect any table content to contain any newlines at this point.  similar to the `float`s, I'd rather remove this logic now and only add it if/when we need it (i.e. once the requirement for such logic is more defined/concrete)

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:92]: can we change the names of the functions `write*` to `*String`?  the  verb `write` makes me think we're actually working with a `Writer` here, as opposed to just string building

**Review by makesitgo — COMMENTED** [review]
ok, this should be the last batch of big comments for the `table.go` file.  wanted to get these in and i'll move on to the test file, but feel free to start addressing these

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:21]: Should I replace **all** references to data into value??

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:54]: Since we need to go through all the values for each header, does it matter if we go through them in the header order? Is there a performance difference with looping through a map vs an array?

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:119]: [naming-nit] I think nick has kind of already covered this, but naming these all `writeX` seems a bit odd since they aren't writing anything, they are just building strings. Maybe `printX` would be better for each since it is printing X and goes well with naming this function `Print`?  `Print` to me makes the most sense since go itself uses that naming convention when printing strings.

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go:21]: no, just within this one function here

File: internal/terminal/table_test.go
Comment (mburdette) [inline internal/terminal/table_test.go:82]: [nit] I think we typically avoid `_` in Test names

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go:54]: what this ensures is our `normalizedData` will be stripped of any row-cell values that are not specified by any of the headers.

i'm thinking of a situation like:

```go
table := newTable(
  []string{"a", "b", "c"},
  []map[string]interface{}{
    "a": "first",
    "b": "second",
    "c": "third",
    "d": "will not show up in Payload() now",
  },
)
```

File: internal/terminal/table_test.go
Comment (mburdette) [inline internal/terminal/table_test.go:102]: For all the occurrences of `"data"` and `"header"` wouldn't it be better to use the `const` values in `table.go` in the tests?

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:123]: Why is this called `Payload` ?  I feel like the term `Payload` is usually associate with api requests. Also can we get comments on exported methods similar to what we do for the server code? 

Also the return type of this seems odd, it just always returns the 2 keys for the data? This doesn't really seem necessary and it seems like we could just have this return `map[string]interface{}, error` Or better yet, wouldn't it make more sense to just have 2 separate functions `Headers()` and `Data()` that basically act as getters?

Maybe I'm missing something here though...

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:112]: Should this method be called `Print` since it is printing the table? Also can we get some comments on exported methods similar to what we do in the server code (Usually the linter picks this up there)?

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:76]: [nit] rename to `printRows`

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:67]: [nit] rename to `printHeader`

File: internal/terminal/table.go
Comment (mburdette) [inline internal/terminal/table.go:92]: [nit] rename to `printDivider`

**Review by mburdette — CHANGES_REQUESTED** [review]
Looking pretty good, mostly naming changes from me

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go:123]: `Message` and `Payload` are interface methods of `LogData`, which supports representing arbitrary data in either string form (`Message`) or JSON (`Payload`).  the return type of `Payload` is the way it is so we can ensure a flat `Log` JSON payload with deterministic key ordering

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go:22]: yeah, same thing. a habit i picked up from eric, from I tend to prefer the `var` zero-value declaration if appropriate (I recently learned, however, Tim does not 🤔 )

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go:123]: linter's probably not complaining b/c its an unexported type.  i think no comments is fine here, or otherwise a comments overhaul is in order b/c all the other types should be doing it too

I'd prefer we just document `LogData` interface really well, instead of docstrings here

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:92]: @mburdette  should it be headerString or printHeader ?

File: internal/terminal/table.go
Comment (jo-tse-mongo) [reply internal/terminal/table.go:77]: If I make the changes to use `make`, what happens is that rows that are empty are defaulted to `""`, so in the end, it adds a newline for each empty row. 
I could filter it out at the end, but it'd be another iteration as opposed to appending it using the `var` method

File: internal/terminal/table_test.go
Comment (jo-tse-mongo) [reply internal/terminal/table_test.go:102]: Most of these occurrences of "data" and "header" are really values within the actual header that just so happened to be named "header" and "data"

I'll change it for the `Payload` test though

File: internal/terminal/table.go
Comment (makesitgo) [reply internal/terminal/table.go]: i think you have it right, empty rows should be omitted.  does this same behavior occur when using `make` with the `0` valued-length? for example: `cells := make([]string, 0, len(t.data))` (as opposed to just `cells := make([]string, len(t.data))`)?

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:93]: sorry about this, i was being dumb earlier!  I didn't connect the dots this was stringifying all the rows, the name should probably either `rowsString` or `dataString` (I think I'm leaning towards the latter, but your call there).

and then `cells` should probably turn into `rows` instead.... sorry about that :bushes:

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:98]: then `row` here could be `cells`

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:113]: can this name be something pluralized, since its an array?  even `parts` would be fine, or `cells` to remain consistent with the other `*String` methods

File: internal/terminal/table.go
Comment (makesitgo) [inline internal/terminal/table.go:130]: [nit] `// leave zero-value` or `// do nothing`

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:10]: I think we can combine the `TestEmptyTableConstructor` and `TestTableConstructor` into a single `TestNewTable` test function.

the test table would then look like:

```go
for _, tc := range []struct{
  description string
  header []string
  data []map[string]interface{}
  expectedData []map[string]string
  expectedWidths []map[string]int
}{
  // test cases
} {
  // inside the test here, you could then just assert.Equal(t, tc.expectedData, t.data) and the like
}
```

what this will allow is:
1. the two empty constructor tests you have here are effectively equal, then they could be defined with the above test case struct as simply:

```go
{
  description: "Should return an empty table if no headers are provided",
}
```

(no declarations for `tc.header` nor `tc.data` are needed here, and the equality checks with `expectedData` and `expectedWidths` also defined as `nil` should pass just fine)

the other two test cases you have for `TestTableConstructor` should fit in nicely here as well

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:78]: with enough test cases written for `TestNewTable`, then the test for `TestTableMessage` and `TestTablePayload` should end up being very simple, we'll likely only want to confirm:
1. with normal data, normal output is produced (edge case-y tests should likely all live in `TestNewTable`)
1. with `nil` headers, we return an error

I see that's basically what you already have with `TestTablePayload`, so lets just port the edge cases from `TestTableMessage` here to `TestNewTable`

**Review by makesitgo — COMMENTED** [review]
I think its nearly there! just this last bit, and I'll wait to see how the test file shakes out before finally approving here.

again, really nice work here.  i'm loving how this is all coming together!

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:22]: can remove `data` here, it's irrelevant to the test path

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:29]: can remove `data` here, it's irrelevant to the test path

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:154]: this test case could be accomplished with an assert on `expectedColumnWidths` in `TestNewTable` above (and will prove to be a little easier to read without all the string formatting required in the message assertion here)

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:128]: this test case could be accomplished with an assert on `expectedData` in `TestNewTable` above (and will prove to be a little easier to read without all the string formatting required in the message assertion here)

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:222]: the error and success cases should be separate `t.Run`s, instead of using conditional asserts here

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:171]: this test case should be duplicated somehow, but with `color.DisableColor = true` set at the start of one of the tests. then we can assert the string is printed without bold formatting as well

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:94]: this test case could be accomplished with an assert on `expectedData` in `TestNewTable` above (and will prove to be a little easier to read without all the string formatting required in the message assertion here)

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:108]: this test case could be accomplished with an assert on `expectedData` in `TestNewTable` above (and will prove to be a little easier to read without all the string formatting required in the message assertion here)

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:234]: you can omit `data` here, its not part of the test path

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:238]: this is fine as is, just commenting here to say for error tests, it's totally ok to not run assertions on the returned values other than `err`

doing `_, _, err := table.Payload()` would be acceptable here as well

File: internal/terminal/table_test.go
Comment (makesitgo) [inline internal/terminal/table_test.go:255]: does this still want to be named `dataStr`?  I think you could go `expectedData` here, or just `data` (which would necessitate you to inline `header` and `data` above into the `newTable` constructor)

**Review by makesitgo — APPROVED** [review]
just a few last cleanup items/questions, but it will be good to go after that

let's just sync up prior to merging here so we can quickly touch base on the rebase/squash/commit message protocol here

nice work!

🆒  beans

File: internal/terminal/table_test.go
Comment (jo-tse-mongo) [reply internal/terminal/table_test.go:154]: For these tests, I left it mostly to check for the whitespace after the values as well as to see what the result of arrays, maps, etc. look like in this table printing format

**Review by makesitgo — APPROVED** [review]
🆒  beans

**Comment by mburdette** [comment]
@jo-tse-mongo the PR title and main commit don't need `realm-cli` in them, since we know repo we are in 😅

File: internal/terminal/table.go
Comment (mburdette) [reply internal/terminal/table.go]: the string print methods in golang use `sprint` so maybe that makes sense ?
https://golang.org/pkg/fmt/#Sprint

**Review by mburdette — APPROVED** [review]
LGTM

**Pull request merged**
